Chez Scheme: add suport for lists in cptypes

gus-massa opened the pull request:
With `list-assuming-immutable?` and the internal construct `immutable-list` the compiler can assume that some lists will
not be mutated.

This is used only to remove some `list?` checks.

mflatt commented:
This change has Racket-level implications, right? If so, consider adding a test at that level now that "optimize.rktl" is back.

gus-massa commented:
I fixed a typo in the description of the first commit, but it's unchanged.

I uncommented a few test and added two additional checks in `test-implies` so when it test that 

    null? => list?

it will also test

    null? => null?
    list? => list?

I don't want to add this automatic test to `test-predicate` in the test block above, because many of the tests are very difficult to track or not yet tracked in CS. 

Anyway, I got an error because it `k:interned-char?` is not tracked (and difficult to track). Anyway^2, I think that all the chars are interned, and `k:interned-char?` should be a platform specific function, so I "fixed" it.

Anyway^3, I get another error, because now `k:interned-char?` is equivalent to `char?`, so `test-implies`
 complains that 

    char? => k:interned-char?

but this test is easy to be removed, if this is the correct behavior in CS.

gus-massa commented:
I squashed everything, because I had too many small interleaved fixups.

I added an special case for `cdr` so

    (lambda (x)
      (when (list-assuming-immutable? x)
        (list? (cdr x))  ;==> #t
      ))

mflatt commented on racket/src/ChezScheme/mats/cptypes.ms:
```diff
@@ -584,6 +597,9 @@
   (not (test-disjoint* '(list? pair?)))
+  (not (test-disjoint* '(list-assume-immutable? null?)))
+  (not (test-disjoint* '(list-assume-immutable? pair?)))
+  (not (test-disjoint* '(list-assume-immutable? list?)))
```
Should these be `list-asuming-immutable?` (with "ing")?

mflatt commented on pkgs/racket-test-core/tests/racket/basic.rktl:
```diff
@@ -540,7 +540,7 @@
   (test #t k:interned-char? #\ )
   (test #t k:interned-char? '#\newline)
-  (test #f k:interned-char? #\u100)
+  (test (not (eq? 'chez-scheme (system-type 'vm))) k:interned-char? #\u100)
```
Looks like this should be without `not`.

mflatt commented:
Looks good, aside from small test corrections.
